### DISCUSSION TITLE
Generalizing from vector to a collection of arrays and, generally, any shape of the scalar container

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#6] [Ideas]
Edit: See also https://github.com/Mikolaj/horde-ad/issues/4.

--

Edit: it seems I'm mixing up this problem (general shape of inputs) and the task of adding arrays as primitive differentiable types (see #11). I think we should first tackle array primitives because our benchmarks show that handling each R scalar in isolation is orders of magnitude slower than treating array operations and their gradients as primitives. I bet that's the case even with selective fusion (assuming no extension to the `Delta` type). Perhaps once we add arrays, the problem in this here discussion will turn out much easier, e.g., due to lowered performance expectations.

--

Tl;dr: The user should be able to keep input parameters (the domain of the function to differentiate) in tuples of unboxed arrays. To enable this without degrading our current peformance, we need to generalize the `eval` machinery from the paper from `DeltaMap` implemented as an unboxed vector to such arbitrary tuples of unboxed arrays. Additionally, the user should be able to keep intermediate and final results not in arrays of dual numbers, but in pairs of unboxed arrays of scalars and corresponding boxed arrays of deltas. This is hard to do generally and efficiently.

Full version: Currently we keep all neural net parameters in one big unboxed vector and we index at various offsets when we use the parameters when building consecutive layers of the network (routinely making indexing mistakes). It would improve readability and correctness of neural network if we could instead keep the parameters in a tuple of (dependently typed) unboxed arrays, one per neural network level. 

Additionally, array operations are, for various reasons, faster than their emulation that work iteratively over individual elements, as benchmarks elsewhere in this repo indicate. Using proper array arithmetic operations would speed up both the computation of the value of the function and of its gradient. To enable the latter, we should selectively fuse (or, manually write) the definitions of gradients for array operations, making all stages of the differentiation take significantly less time and memory. Moreover, for efficiency of both value and gradient computation, we need to work on pairs of unboxed arrays of scalars and corresponding boxed arrays of deltas, which is much harder to set up thatnthe natural boxed arrays of pairs, that is, dual numbers.

In general, we'd like the flexibility to use many different shapes (but not necessarily all that are common in Haskell) of scalar containers as the domain of the function to be differentiated. At the same time, keeping all scalars in unboxed form has proven beneficial for avoiding memory leaks and it makes both computing the value of the function and its gradient fast, as benchmarks indicate. For this reason, e.g., Traversable shapes are out of the question, unless there are special tricks or variants to do so. If there a conflict between the unboxed representation and convenience or safety, perhaps we could delegate the non-zero-cost abstractions to user-friendly wrappers of our library, keeping the kernel of the library lean and efficient.

**Comment by Mikolaj** [comment]
Some suggestions to date:

1. "I wonder about starting with a translation of the form D{Vec R} -> Vec R x Vec Delta, as in section 8.3?
Then one might generalize to containers mapping an arbitrary index set to values, perhaps using the hTensor library?"

1. "Section 11 (section 9 in the new version) of the paper "Generalising to arbitrary S and T" should be a good starting point.
.
It is based on the idea of identifying all the Floats in S and T.  That is not easy to do in Haskell.  I think we want to work with
    f :: S Float -> T Float
where S and T are polymorphic data structures.  For example S might be Vector, or List.  Or
   data S a = MkS (Vector a) (a,a)
which specifies that f takes a vector of Floats and a pair of Floats.
.
Now the "reassociated" and "zip"  operations described in the paper should be do-able with (I hope) standard type classes like Functor and Traversable."

1. "I believe generic-lens takes a very solid stab at this. You might find some of the ideas there helpful if you want to avoid the parameterization which Simon suggests."

1. [archived] I wonder if we could think of the containers as arbitrary
manifolds that however all have unboxed vectors of their elements
underneath, accessed via an explicit coordinate system, in the spirit of
https://hackage.haskell.org/package/goal-geometry
The price, if it works at all, is that we'd need to carry around the explicit
mapping (the coordinate system) from our manifold to the underlying vectors.
If the mapping is non-trivial, it would incur some constant cost.
If hmatrix is faster than unboxed vectors on some operations,
that makes this approach problematic, as well, unless our coordinate
system can replicate the hmatrix smarts (unless they are in C?).

1. MonoTraversable (https://github.com/snoyberg/mono-traversable/tree/master/mono-traversable#readme)

1.  ConstraintKinds, which for some reason has fallen out of favour: https://github.com/mikeizbicki/subhask https://hackage.haskell.org/package/constrained

**Comment by Mikolaj** [comment]
> 1. "I wonder about starting with a translation of the form D{Vec R} -> Vec R x Vec Delta, as in section 8.3?
>        Then one might generalize to containers mapping an arbitrary index set to values, perhaps using the hTensor library?"

This vector of pairs to pair of vectors translation is already performed for the vector that is the domain of the function (the vector of variable parameters). Otherwise, we'd not have the performance we have, especially in benchmarks involving gradient descent. OTOH, in benchmarks with a single gradient computation, such as `prod`, this already causes less readable and a bit less performant code than a vector of pairs would. So perhaps we need the reverse translation or an alternative API to the engine. However, this is only for the domain of the function. In the codomain, e.g., the 10-element vector classifying digits in the MNIST neural network, we have a vector of pairs. Let's experiment how a pair of vectors would would influence the ease of use and the performance.

Edit: I've naively switched all boxed vectors of dual numbers (both the 10-element result and all intermediate records, because only the input parameters vector was covered so far) to pairs of unboxed vector of R and a boxed vector of Delta. This is on branch https://github.com/Mikolaj/mostly-harmless/tree/pairOfV and works, but is slower despite a better representation, due to a lot of temporary vector allocation in the naive conversion code. Unfortunately, it's technically difficult to unzip a vector to two different kinds of vectors and especially hard to efficiently implement operations like `mapM` or `generateM` that create the vector pairs using monadic operations. For starters, I'd need `IO` underneath our monad and the gradient calculation to result in IO. Also, there would be a lot of hand-crafting of each vector operation that needs the pair-of-vectors counterpart and it's hard to imagine how that generalizes to other scalar container shapes.

**Comment by Mikolaj** [comment]
> 2. "Section 11 (section 9 in the new version) of the paper "Generalising to arbitrary S and T" should be a good starting point.

Yes, this makes sense, but the problem is how to do that for, e.g., arbitrary tuples of unboxed arrays.

> Now the "reassociated" and "zip" operations described in the paper should be do-able with (I hope) standard type classes like Functor and Traversable."

Sadly, this doesn't work for unboxed stuff, due to the extra `Unbox` constraint that prevents `Functor` or `Traversable` instances. Most of the other points in this discussion try to work around this technical problem.

**Comment by Mikolaj** [comment]
> 3. "I believe generic-lens takes a very solid stab at this. You might find some of the ideas there helpful if you want to avoid the parameterization which Simon suggests."

I'm not particularly familiar with `lens`, so I'm slowly trying to wrap my head around how this would let me treat a tuple of unboxed arrays in a uniform way, regardless of how many elements this tuple has and what rank and, optionally, what type-expressed size the arrays have (if we use arrays with their size contained in their type, such as https://hackage.haskell.org/package/hmatrix-0.20.2/docs/Numeric-LinearAlgebra-Static.html).

The packages in question: https://hackage.haskell.org/package/generic-lens https://hackage.haskell.org/package/generic-optics https://hackage.haskell.org/package/generic-lens-core

**Comment by Mikolaj** [comment]
> 4. I wonder if we could think of the containers as arbitrary manifolds that however all have unboxed vectors of their elements underneath, accessed via an explicit coordinate system

I think this one doesn't work, because our benchmarks have since shown that hmatrix on matrices is much faster than its element-wise emulation on a single unboxed vector. That's in part, because it uses C matrix libraries, which means we can't emulate the smart approach in our Haskell code, without changing the single unboxed vector representation.

**Comment by Mikolaj** [comment]
> 5. MonoTraversable (https://github.com/snoyberg/mono-traversable/tree/master/mono-traversable#readme)

I think a tuple of unboxed arrays could already be given a MonoTraversable instance. The question remains how to implement effcient `eval` with that. Converting to lists and back is not an option. Actually, foregoing updates in place would already cost us, possibly, half the performance.

Another question is how this helps with systematically implementing operations on arrays of dual numbers as operations on pairs of unboxed and boxed arrays (not at all?).

**Comment by Mikolaj** [comment]
> 6. ConstraintKinds, which for some reason has fallen out of favour: https://github.com/mikeizbicki/subhask https://hackage.haskell.org/package/constrained

I'd love to understand why it's no longer hot. E.g., what's the advantage of `MonoTraversable` that apparently accomplishes the same goal ("Traversable" for unboxed vectors). Is it just that `MonoTraversable` seem simpler notationally? Is type-checking easier, too? Any advantages of `ConstraintKinds` that might be relevant to us? I bet there are some blog posts newer than from the hot period a decade ago.

**Comment by Mikolaj** [comment]
Here's an illustration that the current experiments about implementing vector and matrix (and tensor) operations inform but do not determine the solution to the other problem, namely how and how freely the user can define the shapes of containers of scalars that comprise the domain of the function that the user wants to differentiate. In fact, the two problems are quire intermingled and also related to the problem what codomain types are permitted (currently, these are only the primitive differentiable types, but note that they are now closed under taking a vector of (up to isomorphism), so R^n is permitted).

As per the provisional implementation now, the domain of the function to be differentiated is a tuple of a (unboxed) vector of scalars, a (boxed) vector of vectors and a vector of matrices (and a vector of arbitrary tensors, shortly). In other words, this a tuple of vectors of all our primitive differentiable types. As implemented, this also restricts the possible result types of `Delta` constructors to be the primitive differentiable types (because evaluating a delta expression of a given type results in updating (in a vector) the corresponding partial derivative cell of this type).

While this seems quite sufficient in practice, it's not general by any stretch and it also makes some operations less performant that they could be. For example the following dual number operation

```Haskell
    splitAt1 :: Int -> DualNumber (Vector r) -> DualNumber ((Vector r, Vector r))
```

would require the following `Delta` constructor to be efficiently implemented

```Haskell
    SplitAt1 :: Int -> Delta (Vector r) -> Int -> Delta ((Vector r, Vector r))
```

but this constructor should not be added, because cells of type `(Vector r, Vector r)` are not possible in the domain and so the constructor could not be evaluated.

Therefore, instead of using the splitting constructor `SplitAt1`, we need to construct a delta-expression of result type `Delta ((Vector r, Vector r))` using two copies of the already implemented slice constructors `Slice1`. In terms of the primal component of dual numbers, there is no performance difference (because these are vectors; on lists there would be a difference). However, in terms of delta expressions, there is a significant difference. Each of the two slice constructors allocates a vector of size `k` and makes a call to `evalV`

```Haskell
    Slice1 i n d k -> evalV (konst 0 i V.++ r V.++ konst 0 (k - i - n)) d
```

while `SplitAt1`, evaluated in the context of a pair of vectors `(r1, r2)` of correct lengths, would allocate only one vector of size `k` and would make only one call

```Haskell
    SplitAt1 i d k -> evalV (r1 V.++ r2) d
```

If there's enough payoff for a particular use case, a pair of vectors of the same type could be represented by two slices summing to one vector and then `SplitAt1` allocates no vectors at all. This could even be implemented within the current restrictions on the domain of the differentiated function. But the general case of tuples of different types requires lifting the restrictions.

I have no idea how hard such an extension would be, It would probably entail adding cells of various types (e.g., tuples) that are potentially not present in the actual domain of the function to be differentiated, similarly as currently cells are added for each `Var` constructor regardless if the `Var` comes from domain or from delta-let. If we end up implementing such an extension wrt the actual domain of the differentiated function, to provide the user with a more general API, it would also give us `SplitAt1` for free.
